### PR TITLE
Create ProbeFeatureSelection class v2

### LINKED ISSUE #292: feat: Random Noise Selection method
**Is your feature request related to a problem? Please describe.**
Filtering Noisy features from a set of features can be easily accomplished by adding a one or more random variables to train set, train a gradient boosting or linear model and compute the feature importance.

When a feature is  below the importance of the random noise probably that feature is not so important.

[Reference](https://www.youtube.com/watch?v=RtqtM1UJfZc&t=3150s)

**Describe the solution you'd like**

A `NoiseFilterSelection` method that trains a gradient boosting model (sklearn is a required dependency already) and returns only the features with importance above the random variable.

### PR DESCRIPTION
**Pull request by Morgan-Sell**
Creating new PR for the ProbeFeatusreSelection class due to complications when rebasing.

Closes #292 and #573.

### CODE REVIEW

**Comment by solegalli** [comment]
Hey @Morgan-Sell 

GIthub tells me that the tests were passing in the branch that I pushed. How is it that they are not passing now? Something broke somewhere down the line.

**Comment by Morgan-Sell** [comment]
@solegalli, yeah, sorry, I'm working on it.

As an FYI, I have a ton going on with the house remodeling. My home is a construction zone and I'm playing general contractor.

File: docs/user_guide/selection/ProbeFeatureSelection.rst [hunk 1]
```diff
@@ -0,0 +1,224 @@
+.. _probe_features:
+
+.. py:currentmodule:: feature_engine.selection
+
+ProbeFeatureSelection
+=====================
+
+ProbeFeatureSelection() generates one or more probe features based on the
```
Comment (solegalli) [inline docs/user_guide/selection/ProbeFeatureSelection.rst:8]: We need to wrap the class like these: :class:`ProbeFeatureSelection()` so that in the documentation, it links to the API part.

Comment (solegalli) [reply docs/user_guide/selection/ProbeFeatureSelection.rst:8]: I'd also add "generates one or more random features..."

File: docs/user_guide/selection/ProbeFeatureSelection.rst [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+randomly generated variable, i.e., probe feature, inherently possesses a high level of
+noise. Consequently, any variable that demonstrates less importance than a probe feature
+is assumed to be noise and can be discarded from the dataset.
+
+When initiating the ProbeFeatureSelection() class, the user has the option of selecting
+which distribution is to be assumed to create the probe feature(s) and the number of
+probe features to be created. The possible distributions are 'normal', 'binary', 'uniform',
+or 'all'. 'all' creates 1 or more probe features comprised of each distribution type,
+i.e., normal, binomial, and uniform.
+
```
Comment (solegalli) [inline docs/user_guide/selection/ProbeFeatureSelection.rst:30]: excellent text thank you!

File: docs/user_guide/selection/ProbeFeatureSelection.rst [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+Let's import the required libraries and classes:
+
+.. code:: python
+
+    import pandas as pd
+    import numpy as np
+    from sklearn.ensemble import RandomForestClassifier
+    from sklearn.model_selection import train_test_split
+    from feature_engine.selection import SelectByInformationValue
```
Comment (solegalli) [inline docs/user_guide/selection/ProbeFeatureSelection.rst:50]: I think here we need to import ProbeFeatureSelection

File: docs/user_guide/selection/ProbeFeatureSelection.rst [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+    ((455, 30), (114, 30))
+
+
+Now, we set up :class:`ProbeFeatureSelection()`. We will pass  `RandomForestClassifier()`
+as the :code:`estimator`. We will use `precision` as the :code:`scoring` parameter
+and `5` as :code:`cv` parameter, both parameters to be be used in the cross validation.
+We will assume `1` for the :code:`n_probes` parameter and `normal` as the :code:`distribution`,
+both parameters to be used when creating the probe feature.
+
```
Comment (solegalli) [inline docs/user_guide/selection/ProbeFeatureSelection.rst:100]: Awesome! I did not know you could quote the docstrings using :code:! Thank you for the teachings :)))))

File: docs/user_guide/selection/ProbeFeatureSelection.rst [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    )
+
+    sel.fit(X_train, y_train)
+
+With :code:`fit()`, the transformer:
+
+    - creates `n_probes` number of probe features using provided distribution(s)
+    - uses cross-validation to fit the provided estimator
+    - calculates the feature importance score for each variable, including probe features
+    - if there are multiple probe features, calculate the average importance score
```
Comment (solegalli) [inline docs/user_guide/selection/ProbeFeatureSelection.rst:122]: should it be calculates with s?

File: docs/user_guide/selection/ProbeFeatureSelection.rst [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    sel.features_to_drop_
+
+    ['mean symmetry',
+     'mean fractal dimension',
+     'texture error',
+     'smoothness error',
+     'concave points error',
+     'fractal dimension error']
+
+We see that the :code:`features_to_drop` have feature importance scores that are less
```
Comment (solegalli) [inline docs/user_guide/selection/ProbeFeatureSelection.rst:165]: features_to_drop_ is missing a `_`

File: docs/user_guide/selection/ProbeFeatureSelection.rst [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+     'worst texture',
+     'worst perimeter',
+     'worst area',
+     'worst smoothness',
+     'worst compactness',
+     'worst concavity',
+     'worst concave points',
+     'worst symmetry',
+     'worst fractal dimension']
+
```
Comment (solegalli) [inline docs/user_guide/selection/ProbeFeatureSelection.rst:223]: @Morgan-Sell this document is just great! Thank you so much.

If I get to be greedy, I could I ask you to add a few more snippets of code to create 3 probe features with the param `"all"` and then plot histograms of those features to show the users their distribution?

File: feature_engine/selection/probe_feature_selection.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    n_probes: int, default=1
+        Number of probe features to be created. If distribution is 'all',
+        n_probes must be a multiple of 3.
+
+    distribution: str, default='normal'
+        The distribution used to create the probe features. The options are
+        'normal', 'binomial', 'uniform', and 'all'. 'all' creates at least 1 or more
+        probe features comprised of each distribution type, i.e., normal, binomial,
+        and uniform. The remaining options create `n_probes` features of the selected
+        distribution
```
Comment (solegalli) [inline feature_engine/selection/probe_feature_selection.py:88]: final stop here is missing

File: feature_engine/selection/probe_feature_selection.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+    def fit(self, X: pd.DataFrame, y: pd.Series):
+        """
+        Find the important features.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+
+        y: array-like of shape (n_samples)
+                   Target variable. Required to train the estimator.
```
Comment (solegalli) [inline feature_engine/selection/probe_feature_selection.py:163]: I think this has extra indentation and may cause problems when rebuilding the docs.

File: tests/test_selection/test_probe_feature_selection.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+        scoring="neg_mean_squared_error",
+        n_probes=6,
+        distribution="all",
+        cv=3,
+        random_state=1,
+        confirm_variables=False
+    )
+
+
+    sel.fit(X, y.values.ravel())
```
Comment (solegalli) [inline tests/test_selection/test_probe_feature_selection.py:130]: This is doing what is expected and we can test the generation of the probe features, but there are a lot of other things happening as well. 

My idea was to call the method `_genearate_probe_features()` with the required parameters, it returns a df, capture the df in a variable and then assert if equal to the expected.

File: tests/test_selection/test_probe_feature_selection.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+    expected_results = {
+        "var_2": [0.062, -0.051, 0.044, -0.012, -0.036],
+        "var_3": [0.022, -0.026, -0.006, -0.037, 0.022],
+        "var_8": [0.02, -0.068, 0.003, 0.023, -0.032],
+    }
+    expected_results_df = pd.DataFrame(expected_results)
+
+    assert results.head().round(3).equals(expected_results_df)
+
+# def test_transformer_with_binary_distribution(load_diabetes_dataset):
```
Comment (solegalli) [inline tests/test_selection/test_probe_feature_selection.py:174]: fixing this is wip?

**Comment by solegalli** [comment]
This is looking really good @Morgan-Sell 

I left a few comments, but aside the one for the new test, all others are minor details.

Give me a shout when you need me to take another look.

Thank you so much!

File: docs/user_guide/selection/ProbeFeatureSelection.rst (hunk 7 above)
Comment (Morgan-Sell) [reply docs/user_guide/selection/ProbeFeatureSelection.rst:223]: How can I say no to profesora che? ;)

**Comment by Morgan-Sell** [comment]
hola @solegalli,

We're getting there! I need to run isort/black and fix some formatting/style bugs. I'll add the histograms too.

Once I'm finished, I'll review all the text.

Is there anything else?

**Comment by solegalli** [comment]
Hey @Morgan-Sell 

Awesome! Looking forward to the final version.

Did you add it to this test file? https://github.com/feature-engine/feature_engine/blob/main/tests/test_selection/test_check_estimator_selectors.py

Once there, we hope that the tests pass, and we are done. 💪🏽

**Comment by Morgan-Sell** [comment]
hi @solegalli ,

I need to see why the tests are now failing.

I'm struggling w/ making a reasonable-looking histogram with an example of the 3 distributions. Do you have an example?
